text/0436-gamelift-l2.md

```diff
@@ -0,0 +1,983 @@
+  // cdk.Taggable: IMatchmakingConfiguration allows tagging
+  interface IMatchmakingConfiguration extends cdk.IResource, iam.Grantable, cdk.ITaggable {
+    readonly matchmakingConfigurationArn: string;
+    readonly matchmakingConfigurationName: string;
+    grant(grantee: iam.IGrantable, ...actions: string[]): iam.Grant;
+    grantStartMatchmaking(grantee: iam.IGrantable): iam.Grant;
+    metric(metricName: string, props?: cloudwatch.MetricOptions): cloudwatch.Metric;
+    // Some canned metrics as well like `metricMatchAccepted`
+  }
+```
+
+* `MatchmakingConfiguration` -- matchmaking configuration class with some helper props/methods to help build matchmaking ruleset and other common configuration.
+
+```ts
+enum MatchmakingMode { STANDALONE, WITH_QUEUE}
+enum BackfillMode { MANUAL, AUTOMATIC}
+interface MatchmakingProps {
+  readonly name?: string;
+  // The ruleSet used to definied matchmaking conditions
+  readonly ruleSet: IRuleSet;
+  readonly backfillMode?: BackfillMode;
+  readonly queues?: IQueue[];
+  readonly mode?: MatchmakingMode;
+}
+
+abstract class MatchmakingConfiguration implements IMatchmaking {
+  constructor(protected readonly props: MatchmakingConfigurationProps = {}) {}
+  // Helper methods that subclasses can use to create common config
+  protected createGameProperty(...): CfnMatchmakingConfiguration.GamePropertyProperty | undefined;
+  protected createRuleSet(...): CfnMatchmakingRuleSet | undefined;
+  protected createQueue(...):
+  CfnGameSessionQueue | undefined;
+}
+```
+
+* `IMatchmakingRuleSet` -- interface to define and configure matchmaking ruleSet and produce configuration that is injected into the
+  Matchmaking definition.
+
+```ts
+// cdk.IResource: Since IMatchmakingRuleSet will extend Resource
+// iam.Grantable: To allow service role to access other resources like GameLift Fleet or Game session queue or other ressources
+// cdk.Taggable: IMatchmakingRuleSet allows tagging
+interface IMatchmakingRuleSet extends cdk.IResource, cdk.ITaggable {
+  readonly ruleSetArn: string;
+  readonly ruleSetName: string;
+  abstract bind(scope: Construct, options: RuleSetBindOptions): RuleSetConfig;
+  metric(metricName: string, props?: cloudwatch.MetricOptions): cloudwatch.Metric;
+  // Some canned metrics as well like `metricRuleEvaluationsPassed`
+}
+```
+
+* `MatchmakingRuleSet` -- absdtract base matchmaking class with some helper props/methods to help build matchmaking ruleset and other common configuration
+
+```ts
+enum PlayerAttributeType {
+    STRING = 'string',
+    NUMBER = 'number,
+    STRING_LIST = 'string_list',
+    STRING_NUMBER_MAP = 'string_number_map'
+}
+interface TeamProperty {
+  readonly name: string;
+  readonly maxPlayers: number;
+  readonly minPlayers: number;
+  readonly quantity: number;
+}
+
+interface PlayerAttributeProperty {
+  readonly name: string;
+  readonly type: PlayerAttributeType;
+  readonly default: string;
+}
+
+interface AlgorithmProperty {
+  readonly strategy: string;
```

Comment:
Is this a closed set? If so, it could benefit from an enum.

Reply:
unfortunately not, there is a wide range of possibilities here:
"algorithm": {
        "strategy": "balanced",
        "batchingPreference": <"largestPopulation", "fastestRegion">,
        "balancedAttribute": "string",
        "expansionAgeSelection": <"newest", "oldest">,
        "backfillPriority": <"normal", "low", "high">
    },

Reply:
I mean the `strategy` property specifically.

Reply:
For now it's unfortunately a free text on the MatchmakingRuleSet schema.

Reply:
Done